### DISCUSSION TITLE
Cariddi v2

### DISCUSSION THREAD

**Discussion by edoardottt** [edoardottt/cariddi#94] [Ideas]
This discussion will be used to share ideas / roadmap for cariddi v2.  
cc @cyb3rjerry  

- https://github.com/edoardottt/cariddi/issues/96
- https://github.com/edoardottt/cariddi/issues/95
- https://github.com/edoardottt/cariddi/issues/98
- https://github.com/edoardottt/cariddi/issues/89
- https://github.com/edoardottt/cariddi/issues/100
- https://github.com/edoardottt/cariddi/issues/101

**Comment by humpty-tony** [comment]
:wave: So here would be my initial proposed changes

## Wrap the [New()](https://github.com/edoardottt/cariddi/blob/4d590282a85f9fd2f64f279018e7dbc03c520711/pkg/crawler/colly.go#L55-L60) params into a single object. 
The benefits of this would be better readability, better control over the mutability of the object considering multiple instances of it can be spawned and would be a little more idiomatic from a Go perspective.
The function would then look something more like:
```go
func New(scanCtx *ScanCtx){
...
}
```
Where the ScanCtx struct could look like:
```go
type ScanCtx struct {
	Debug         bool
	ErrorsFlags   bool
	EndpointsFlag bool
	EndpointsFile bool
	InfoFlag      bool
	Insecure      bool
	Plain         bool
	SecretsFlag   bool
	FileType      int
	DelayTime     int
	Threads       int
	Html          string
	Ignore        string
	Target        string
	Txt           string
	SecretsFile   []string
	Headers       map[string]string
}

```

## Wrap the results in a struct that has a [sync.Mutex](https://pkg.go.dev/sync#Mutex) to prevent data races
Similarily to my previous point, we could also return a single object that has a Mutex attribute to make sure we don't cause data races when writing them.

The object could look something like
```go
type Results struct {
	Mutex      sync.Mutex
	Results    []string
	Secrets    []scanner.SecretsMatched
	Endpoints  []scanner.EndpointMatched
	Extensions []scanner.FileTypeMatched
	Errors     []scanner.ErrorMatched
	Infos      []scanner.InfoMatched
}
```

We could then easily write to the object by doing something like:
```go
results.Mutex.Lock()
results.Results = append(results.Results, something...)
results.Mutex.Unlock()
```
A good example of how to use Mutexes to prevent race conditions can be seen [here](https://gobyexample.com/mutexes)

**Reply by humpty-tony** [reply]
Instead of using Mutexes we could also instead pipe all results into a chan that then gets processed. We don't really care for the order of the results, only that there **is** some results. This could be a very clean way of handling concurency

**Comment by edoardottt** [comment]
### Wrap the New() params into a single object.
Absolutely ok for this one.

Regarding the struct, it should contain all the parameters passed to the New function now, something like this:
```Go
type Scan struct {
	Target        string
	Txt           string
	Html          string
	Delay         int
	Concurrency   int
	Ignore        string
	IgnoreTxt     string
	Cache         bool
	Timeout       int
	Intensive     bool
	Rua           bool
	Proxy         string
	Insecure      bool
	SecretsFlag   bool
	SecretsFile   []string
	Plain         bool
	EndpointsFlag bool
	EndpointsFile []string
	FileType      int
	Headers       map[string]string
	ErrorsFlag    bool
	InfoFlag      bool
	Debug         bool
	UserAgent     string
}
```
Moreover, I think that also CreateColly function https://github.com/edoardottt/cariddi/blob/4d590282a85f9fd2f64f279018e7dbc03c520711/pkg/crawler/colly.go#L322
can take that struct as input.

What do you think?

**Reply by humpty-tony** [reply]
I think this sounds reasonable. We might, however, want a "sub-struct" for results so we can add a little more control from a mutability POV
We could do for example
```go
type Results struct {
       Results    []string
       Secrets    []scanner.SecretsMatched
       Endpoints  []scanner.EndpointsMatched
       Extensions []scanner.FileTypeMatched
       Errors     []scanner.ErrorMatched
       Infos      []scanner.InfoMatched
}

type Scan struct {
	Target        string
	Txt           string
	Html          string
	Delay         int
	Concurrency   int
	Ignore        string
	IgnoreTxt     string
	Cache         bool
	Timeout       int
	Intensive     bool
	Rua           bool
	Proxy         string
	Insecure      bool
	SecretsFlag   bool
	Plain         bool
	EndpointsFlag bool
	FileType      int
	Headers       map[string]string
	ErrorsFlag    bool
	InfoFlag      bool
	Debug         bool
	UserAgent     string
        Results       *Results
}
```

This way, the config is set directly within the `Scran` struct, and the actual results are stored within the `Results` attribute

**Comment by edoardottt** [comment]
### Wrap the results in a struct
Absolutely ok also with this one.

**Comment by edoardottt** [comment]
### Prevent data races
It's up to you to decide which method to use, I'm okay both with mutex or channels.  
Maybe with channels could be a lil bit more annoying to remove duplicates (There shouldn't be duplicates...but just to be sure), **BUT** using this method we can write results in output files while the crawler is scanning. Now instead it crawls everything and then it prints to file at the end.
But we should also rewrite the output write functions to be more flexible.

P.S. With millions of URLs crawled, remove duplicates could be a huge bottleneck, we should think carefully on which method to use

**Reply by humpty-tony** [reply]
Yeah the following method is typically what I do since empty structs [have a size of zero](https://go.dev/play/p/PyGYFmPmMt) and hashtables (maps) have a complexity [near O(1)](https://dave.cheney.net/2018/05/29/how-the-go-runtime-implements-maps-efficiently-without-generics#:~:text=Go%E2%80%99s%20map%20is%20a%20hashmap)
```go
var secretsList :=  make(map[string]struct{})

if _, ok := secretsList["newSecret"]; !ok { // If key doesn't exist, append it
    secretsList["newSecrets"] = struct{}{}
}
```

**Reply by edoardottt** [reply]
Okay, seems good. I have to think about this.

Some random thoughts:
  - Remember to fork also the devel branch and create PRs against the cariddi/devel branch.
  - I think even if we write results to file while crawling it's better to print them on cli at the end (not talking about URLs, but secrets, infos...)

For now we can work on the three issues ready and then go back at it :) I will update the first message of the discussion

**Comment by humpty-tony** [comment]
@edoardottt Do you happen to have a specific URL that you used for testing that could be used as a benchmark?

**Reply by edoardottt** [reply]
Actually no... 
I was thinking that we can build a server for testing the tool locally. in this way we know exactly which pages should crawl or not. we can also insert fake secrets, errors ...

**Reply by humpty-tony** [reply]
Sounds good, got a server I could use for this. If you can make the test pages I'll gladly set it up in my k8s cluster so we'll be able to deploy it with certificates, a clean subdomain, etc.. so it can be as realistic as possible

**Reply by edoardottt** [reply]
I'm going to sleep rn :/ tomorrow I will setup everything  :) thanks again for your contribution ❤️

**Reply by humpty-tony** [reply]
Yeah we can do all of this in due time, no rush :)

**Reply by edoardottt** [reply]
We could also use this https://github.com/google/security-crawl-maze. Of course a test page for secrets, infos, etc... can be created as a standalone test :)

**Reply by humpty-tony** [reply]
Yeah I think we can use that for sure, we might want to sprinkle secrets left and right into it however

**Comment by humpty-tony** [comment]
## Improve Regex
I think we might eventually want to add a little more specificity to the Regex rules as they currently tend to return false positives
![image](https://user-images.githubusercontent.com/92945113/202327038-e78a1974-4790-41bd-ab52-f77c8b97cc27.png)
![image](https://user-images.githubusercontent.com/92945113/202327052-f24eb744-9c37-41a6-8911-0c3258aa2377.png)
![image](https://user-images.githubusercontent.com/92945113/202327092-db5f8061-671a-4f46-a30f-0f07314d4f48.png)
![image](https://user-images.githubusercontent.com/92945113/202327128-e0e1169c-853e-45c7-b247-968d302ded10.png)

**Reply by edoardottt** [reply]
absolutely agree.
- https://github.com/edoardottt/cariddi/issues/101

**Comment by humpty-tony** [comment]
## Refactor `visitHTMLLink()`
visitHTMLLink() currently takes 11 params while we could most likely find a way to simply pass the original `Scan` struct and do manipulations within the function

This would turn the 9 `visitHTMLLink(arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9){...}` into `visitHTMLLInk(&Scan)`

**Reply by edoardottt** [reply]
absolutely agree :)
- https://github.com/edoardottt/cariddi/issues/100

**Comment by edoardottt** [comment]
Also CreateColly 
https://github.com/edoardottt/cariddi/blob/devel/pkg/crawler/colly.go#L357 
can take the scan struct as input :) @cyb3rjerry

**Comment by edoardottt** [comment]
@cyb3rjerry I've just updated the devel branch, I've removed the `insecure` flag. Now it skips verifying certificates accepting also invalid ones

**Reply by edoardottt** [reply]
Give a look also to this new tool :P @cyb3rjerry https://github.com/edoardottt/csprecon, let me know if you think it can be improved :)
